source/extensions/filters/network/zookeeper_proxy/zookeeper_utils.cc

```diff
@@ -0,0 +1,53 @@
+#include "extensions/filters/network/zookeeper_proxy/zookeeper_utils.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ZooKeeperProxy {
+
+bool BufferHelper::peekInt32(Buffer::Instance& buffer, uint64_t& offset, int32_t& val) {
+  try {
+    val = buffer.peekBEInt<int32_t>(offset);
+    offset += sizeof(uint32_t);
+    return true;
+  } catch (EnvoyException& e) {
+    return false;
+  }
+}
+
+bool BufferHelper::peekBool(Buffer::Instance& buffer, uint64_t& offset, bool& val) {
+  if (buffer.length() < (offset + 1)) {
+    return false;
+  }
+
+  try {
+    std::string str;
+    str.assign(std::string(static_cast<char*>(buffer.linearize(1 + offset)), 1 + offset));
+    str = str.substr(offset);
+    val = static_cast<bool>(str[0] & 255);
+    offset += 1;
+    return true;
+  } catch (EnvoyException& e) {
+    return false;
+  }
+}
+
+bool BufferHelper::peekString(Buffer::Instance& buffer, uint64_t& offset, std::string& str) {
+  int32_t len = 0;
+  peekInt32(buffer, offset, len);
+
+  if (buffer.length() < (offset + len) || !len) {
+    return false;
+  }
+
+  str.assign(std::string(static_cast<char*>(buffer.linearize(len + offset)), len + offset));
```

Comment:
can you use `copyOut` on the segment you're interested in instead of `linearize` on the entire buffer?

Reply:
Yep, done.